doc/proposals/composable-integration-in sdk.md

```diff
@@ -0,0 +1,173 @@
+
+The Composable SDK has been implemented and open-sourced at https://github.com/IBM/composable/tree/master/sdk.
+It is also related to the Composable Operator released on operatorhub.io:
+https://github.com/IBM/composable.
+
+
+## Motivation
+
+To add support for cross-resource references in CRD definitions and controllers (Composable SDK) and make this
+functionality available through the Operator-SDK.
+
+## Goals
+
+- Provide cross-resource reference types, which can be used in CRD type definitions.
+- Provide resolution functions, which can be called from within a reconciler to resolve cross-resource references.
+
+
+## Proposal
+
+### User Story 1
+
+As an operator developer, I wish not to have the value of a field be hard-wired. These are typically
+values that are computed dynamically when an application is deployed. For example, the admin URL for a Kafka
+deployment is known only after Kafka has been successfully deployed. If my CRD requires the URL, then
+rather that requiring a string value for that field, I can define it to be a reference to another object
+(perhaps the data is in the status field of the Kubernetes object managing Kafka).
+
+### User Story 2
+
+Often the yamls for a collection of resources require the same data in many places (e.g., a host name or port number).
+Rather than hardwiring the same value everywhere, I can define cross-resource references to a unique configmap that contains 
+the data. This makes it easier to change the data if needed, and can allow all resources to be configured
+dynamically (i.e. changes in the configmap are absorbed and resource that point to it are updated).
+
+
+### Implementation Details/Notes/Constraints (optional)
+
+The Composable SDK offers the following types.
+
+```golang
+type ObjectRef struct {
+	GetValueFrom ComposableGetValueFrom `json:"getValueFrom"`
+}
+
+type ComposableGetValueFrom struct {
+	Kind               string   `json:"kind"`
+	APIVersion         string   `json:"apiVersion,omitempty"`
+	Name               string   `json:"name,omitempty"`
+	Labels             []string `json:"labels,omitempty"`
+	Namespace          string   `json:"namespace,omitempty"`
+	Path               string   `json:"path"`
+	FormatTransformers []string `json:"format-transformers,omitempty"`
+}
+```
+
+An `ObjectRef` can be used to specify the type of any field of a CRD definition, allowing the value to be determined dynamically.
+For a detailed explanation of how to specify an object reference according to this schema, see [here](https://github.com/IBM/composable/blob/master/README.md#getvaluefrom-elements).
+
+The Composable SDK offers the following function for resolving the value of cross-resource references.
+
+```golang
+func ResolveObject(r client.Client, config *rest.Config, object interface{}, resolved interface{}, composableNamespace string) *ComposableError 
```

Comment:
The config is used to get a CachedDiscoveryInterface (k8s.io/client-go/discovery), which is used to lookup a resource kind given a kind and an apiversion. This in turn gives information that is used to lookup the reference object.

Reply:
So, it means that it will not respect the [operator scope](https://github.com/operator-framework/operator-sdk/blob/master/doc/operator-scope.md). For example, if we pass just an ns for the manager [here](https://github.com/operator-framework/operator-sdk-samples/blob/master/go/memcached-operator/cmd/manager/main.go#L95) it would be looking for in all ns of the cluster. Am I right? 

If yes, IHMO it is something that needs to be changed.

Reply:
If the object reference have no namespace specified, then it will look only in the namespace of the manager. This gets passed as the composableNamespace parameter to ResolveObject(). That's the default namespace to use.

If the object reference has a namespace specified, it will look there instead.

The Config is used to get a CachedDiscoveryInterface and all we need to know is to see if the kind of the object reference is namespaced or not. This will inform how to create a NamespacedName for object lookup. If it is namespaced, then the convention above is used to determine the namespace.

Reply:
Hi @vazirim, 

Tks for the clarifications. 

IHMO in order to the Compasable work aligned with it would need to use the manager/cache which is provided instead of a new cache. Otherwise, shows that It would allow for example, besides the operator be set up in the manager/cache with a List of namespaces looking for the resource out of this scope. Also, it can have more options setup by default which IHMO should be respected in all project. Is it make sense? 

PS.: The manager can be set up with some options. Regards the ns it can be "" which will allow cache all cluster, or a list of namespaces or just 1 namespace. See: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/manager#Options

Please, let me know if I misunderstood something as well.

Reply:
ResolveObject takes a default namespace at the moment, so this can be set however the user wants to set it. In the Memcached tutorial, it's actually set to be the same as the namespace of the object that contains the reference.

We are however letting objects to have references to objects in other namespaces. I think this is a powerful feature to have. For example, you could imagine one namespace may have some service and some credentials in a secret. Some other service in a different namespace could have a pointer to that secret. IMHO it would be best not to prevent such scenarios.

Reply:
@camilamacedo86 if I am misunderstanding your question, please let me know...

Reply:
Hi @vazirim, 

The tutorial getting started is just the basic steps. However, as described in the doc [operator scope](https://github.com/operator-framework/operator-sdk/blob/master/doc/operator-scope.md) just for the namespace you can pass more options.  

However, indeed other setups can be made in the manager which managers the project and for its cache. 

Because of this IHMO, this part of the proposal and Compasable needs change. It would need to use the same cache. The user can set up where is the resource, however, if it will be out of the scope of the project then it should not be returned.

Reply:
@camilamacedo86 I see what you are saying -- will investigate

Reply:
@camilamacedo86 The CachedDiscoveryClient is actually not an object cache, but rather a cache of Kubernetes supported resource definitions. This discovery interface helps to look up the available resource *kinds*. It does not give access to objects of those kinds.
We use it so that the user can succinctly write the pointers (the Kind field is required, but Apiversion is optional). So the discovery interface helps to look up the precise resource kind when the Apiversion is omitted.
Knowing the resource kind precisely then helps in forming a Get request for the specified object reference. If this object request is incompatible with the Rbac for the operator, then the Get will fail and ResolveObject will return the error.
So the operator is never allowed to reach outside of its scope.
Does it adress your concerns?

Reply:
No. Because the cache which has been used by the why the Compasable is not the same as the operator. It is not about the RBCA. 

IHMO in order to address it the Compasable would need to use the client provided by the controller instead of creating a new cache and client.

Reply:
It is not about the RBCA.
IHMO: To address it the Composable could use the client provided by the controller and not create a new cache.

Reply:
The ResolveObject is indeed using the client provided by the controller:

```func ResolveObject(r client.Client, config *rest.Config, object interface{}, resolved interface{}, composableNamespace string) *ComposableError {```

The CachedDiscoveryInterface is obtained from the controller's client as well (the client passed into the ResolveObject)...

I am not sure I understand the concern (?)

Reply:
Thanks for the explanation @vazirim! I see now why the `rest.Config` is important. But now I have a follow-up. :)

Have you considered making a `ResolvesObject` interface:

For example, 
```go
type ResolvesObject interface {
  // ResolveObject resolves object references. It uses a context
  // for cancellation.
  ResolveObject(ctx context.Context, in, out interface{}) error
}
```

And a struct that implements it:
```go
type KubernetesResourceResolver struct {
  defaultNamespace string
  client client.Client
  resourcesClient discovery.ServerResourcesInterface
}
```

With this, the resolver could be setup and included as a field in the reconciler implementation, it would simplify the user-facing API, and it would make it more extensible for other possible future use cases (and writing unit tests).

Reply:
@joelanford Thanks for the suggestion, I will make the modification.
The only thing is that ResolveObject needs to have a fresh CachedDiscoveryClient at each invocation. This is because new types may have been added between invocations, so the cache can't be defined once and reused for the lifetime of the operator. In this case, I would prefer to have this struct instead:
```
type KubernetesResourceResolver struct {
  defaultNamespace string
  client client.Client
  resourcesClient discovery.CachedDiscoveryClient 
}
```
and then call `invalidate()` on the `resourcesClient` at each invocation.
What do you think?

Reply:
@joelanford The latest version of the Composable SDK (v0.1.4) now has an interface like you have suggested. In addition to the ResolveObject interface, we have the struct that implements it:
```
type KubernetesResourceResolver struct {
	Client          client.Client
	ResourcesClient discovery.ServerResourcesInterface
}
```

We simplified and got rid of the default namespace here. 
The resolveObject now has this interface:
```
func (k KubernetesResourceResolver) ResolveObject(ctx context.Context, object interface{}, resolved interface{}) error 
```
The assumption is that the object passed has a namespace, and that serves as the default interface.

Reply:
> The assumption is that the object passed has a namespace

Forgive me if this has already been answered or addressed. Does this mean that cluster-scoped resources are not supported? Or can I pass an object where namespace is empty and that means cluster-scoped?

Reply:
Cluster-scoped resources are supported as well. We check to see if the reference is to a cluster-scoped resource or not. If not, and if the namespace is omitted from the GetValueFrom reference, then we use the default namespace, which is the namespace of the object itself (that contains the reference).